File: CodeTest/src/Programmers/Lv2/Q5/Q5.py
"""
# Author    : Lee In Bok 
# Date      : 2023.12.30(Sat)
# Spend Time: 09m 27s
# Runtime   : 31.53 ms (Longest Time)
# Memory    : 12 MB (Most Spacious)
# Algoritm  : Dynamic Programming
"""

def solution(n):
    if n == 0 or n == 1:
       return 1

    dp = [0 for _ in range(n + 1)]
    dp[0] = 1
    dp[1] = 1

    for idx in range(2, n + 1):
      dp[idx] = (dp[idx - 2] + dp[idx - 1]) % 1000000007

    return dp[n]

res = solution(10)
print(res)

File: CodeTest/src/Programmers/Lv2/Q5/Q5.java
/**
* Author    : Lee In Bok 
* Date      : 2023.12.30(Sat)
* Spend Time: 09m 27s
* Runtime   : 3.91 ms (Longest Time)
* Memory    : 77.3 MB (Most Spacious)
* Algoritm  : Dynamic Programming
*/

package Programmers.Lv2.Q5;

public class Q5 {
  public static void main(String[] args) {
    Solution s = new Solution();
    int res = s.solution(4);

    System.out.println(res);
  }  
}

class Solution {
    public int solution(int n) {
        int[] dp = new int[n + 1];

        dp[0] = 1;
        dp[1] = 1;

        for(int i = 2; i <= n; i++){
          dp[i] = (dp[i - 2] + dp[i - 1]) % 1000000007;
        }

        return dp[n];
    }
}